### DISCUSSION TITLE
Large batch sizes memory leak in llama.cpp?

### DISCUSSION THREAD

**Discussion by kalomaze** [ggml-org/llama.cpp#5696] [General]
I notice that the larger the batch size, the more memory it requires to do consecutive batches.
This confuses me. Shouldn't the earlier batches not impact the future batches? As in, the memory from the earlier batches should be freed, and it shouldn't escalate beyond the first?

I OOM on the third batch when I set the bs to 2048 tokens:

<img width="355" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/3c1e2470-89d0-4e88-9048-e9d146e763d2">

Is this related to / caused by the fact that Flash Attention is not yet supported?

**Comment by kalomaze** [comment]
![image](https://github.com/ggerganov/llama.cpp/assets/66376113/9f47c10c-faa6-4b3d-9070-d6aaff136c27)
It seems the memory usage keeps crawling continuously for each successive batch, which smells like a memory leak of some kind / memory doesn't get freed somewhere where it should be.

**Reply by kalomaze** [reply]
<img width="345" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/782f23b8-9cf9-439f-9ccd-6564a647b130">

The model launches with 21.6GB. With the first batch of 2048 tokens, It increases to ~22GB. On the batch after that, 22.7GB. 3rd batch, 23.5GB, and then it OOMs when trying to do the next.

On a batch size of 1024, I launch with 19.8GB:

<img width="369" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/7945a005-77ea-488c-b9be-b91fb98ad8e6">

And after 12k tokens, it rises to 21.2GB

Each consecutive batch increases memory in the same way as 2048 batch size does (except it's much less memory, seems to be ~100mb per batch instead of ~500-800mb).
This memory usage does **not** ever go down to where it was when it launched when you reset the context from the start.

Seems like a classic case of a memory leak but maybe this is somehow expected behavior for reasons beyond my understanding, @JohannesGaessler any thoughts?

**Reply by kalomaze** [reply]
With a very small batch size of 64, the model launches with 18GB. On the first batch, this rises to 18.4GB.
After 11,000 tokens, it still reads as 18.4GB. It seems that it slowly grows in memory usage in the same way though, just that it's negligible at this batch size.

It also happens on the CPU.

For batch size 512, every batch increases the non-GPU memory usage by:
17mb
For batch size 1024, it is
65mb
For batch size 2048, it is
242mb

When offloading 18/33 layers.

**Reply by JohannesGaessler** [reply]
How exactly are you running llama.cpp? I'm assuming you're on Windows based on your screenshots but more information is definitely needed here.

**Reply by kalomaze** [reply]
When using main.exe in CLI, I observe this trend on pure CPU, with 0 layers offloaded to GPU.
I built latest llama.cpp with cmake & CuBLAS, as x64-Release.

n_batch 2048 = 256mb increased memory use for each batch
n_batch 1024 = 64mb increased for each batch
n_batch 512 = 16mb increased for each batch

It seems to scale quadratically for whatever reason.

I am sourcing the numbers by comparing the difference in total memory used by main.exe after each consecutive batch according to task manager.

<img width="370" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/473eb530-c4bf-493f-9838-d0eb57a9adc7">

I used a very long test prompt of keymashed numbers for convenience, about ~9k tokens.

`main.exe -m D:/mixtral-8x7b-instruct-v0.1.Q5_K_M.gguf -c 12000 -b 2048 -p "7805453404783859027528703939740684426036520473469503476354354353453453453463046783905843209583409673409574309567039467340967450967450967450698785350934859043760394762049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490780545340478385902752870393974068442603652047346950347632049124031903478693463463456245634590340956734096734067456346309573940673240458320957329057349053470953478988888503925732409567420673096734095674309573409573490574390573490573450973495374509347509345734097609346734907805453404783859027528703939740684426036520473469503476320491240319034786934634634562456345903409567340967340674563463095739406732404583209573290573490534709534789888885039257324095674206730967340956743095734095734905743905734905734509734953745093475093457340976093467349078054534047838590275287039397406844260365204734695034763204912403190347869346346345624563459034095673409673406745634630957394067324045832095732905734905347095347898888850392573240956742067309673409567430957340957349057439057349057345097349537450934750934573409760934673490"`

**Reply by Artefact2** [reply]
You can use `llama-bench -p 12000 -b 2048` fwiw.

Does your GPU use VMM?

**Reply by kalomaze** [reply]
Not sure. If you are asking about passthrough for WSL reasons, I can attempt to test on my Linux partition instead of WSL. But do keep in mind it happens on pure CPU inference as well.

**Reply by Artefact2** [reply]
Offloading 0 layers with cublas isn't the same as pure CPU inference, the kv matmuls are still offloaded, and cublas will allocate internally for those. I think it's the same as #4946

**Comment by JohannesGaessler** [comment]
What I think is happening is that this is one of the side effects of the inefficient attention implementation on master. It should be fixed with FlashAttention, if not I will take another look.
